src/com/linkedin/parseq/Engine.java

```diff
@@ -103,6 +96,37 @@
     _allLogger = loggerFactory.getLogger(LOGGER_BASE + ":all");
     _rootLogger = loggerFactory.getLogger(LOGGER_BASE + ":root");
 
+    if (_properties.containsKey(MAX_RELATIONSHIPS_PER_TRACE)) {
+      _maxRelationshipsPerTrace = (Integer) getProperty(MAX_RELATIONSHIPS_PER_TRACE);
+    } else {
+      _maxRelationshipsPerTrace = DEFUALT_MAX_RELATIONSHIPS_PER_TRACE;
+    }
+
+    if (_properties.containsKey(MAX_CONCURRENT_PLANS)) {
+      _maxConcurrentPlans = (Integer) getProperty(MAX_CONCURRENT_PLANS);
+    } else {
+      _maxConcurrentPlans = DEFUALT_MAX_CONCURRENT_PLANS;
+    }
+    _concurrentPlans = new Semaphore(_maxConcurrentPlans);
```

Comment:
I didn't think of any real life scenario but was just out of curiosity. After a second thought, I think you're right that people expecting FIFO behavior can and probably should submit their tasks one after another using blocking API, so it doesn't matter how semaphore behaves internally.